src/catchup/simulation/ApplyTransactionsWork.cpp

```diff
@@ -35,17 +49,190 @@ ApplyTransactionsWork::ApplyTransactionsWork(
+        }
+    }
+}
+
+bool
+ApplyTransactionsWork::hasSig(PublicKey const& account,
+                              TransactionEnvelope const& oldEnvelope,
+                              Hash const& hash)
+{
+    // Is the signature of this account present in the envelope we're
+    // simulating?
+    if (!mUsedPubKeys.insert(account).second)
+    {
+        return false;
+    }
+
+    for (auto const& sig : oldEnvelope.signatures)
+    {
+        if (PubKeyUtils::verifySig(account, sig.signature, hash))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void
+ApplyTransactionsWork::addSignerKeys(AccountID const& acc,
+                                     AbstractLedgerTxn& ltx,
+                                     std::vector<SecretKey>& keys,
+                                     TransactionEnvelope const& oldEnvelope,
+                                     uint32_t n)
+{
+    auto opKey = SimulationUtils::getNewSecret(acc, n);
+    auto const& txHash = mResultIter->transactionHash;
+
+    if (hasSig(acc, oldEnvelope, txHash))
+    {
+        keys.emplace_back(opKey);
+    }
+
+    auto account = stellar::loadAccount(ltx, acc);
+    if (!account)
+    {
+        return;
+    }
+
+    for (auto const& signer : account.current().data.account().signers)
+    {
+        if (signer.key.type() == SIGNER_KEY_TYPE_ED25519)
+        {
+            auto pubKey = KeyUtils::convertKey<PublicKey>(signer.key);
+            if (hasSig(pubKey, oldEnvelope, txHash))
+            {
+                keys.emplace_back(SimulationUtils::getNewSecret(pubKey, n));
+            }
+        }
+    }
+}
+
+uint32_t
+ApplyTransactionsWork::scaleLedger(
+    std::vector<TransactionEnvelope>& transactions,
+    std::vector<TransactionResultPair>& results,
+    std::vector<UpgradeType>& upgrades, uint32_t n)
+{
+    assert(mTransactionIter != mTransactionHistory.txSet.txs.cend());
+    assert(mResultIter != mResultHistory.txResultSet.results.cend());
+
+    LedgerTxn ltx(mApp.getLedgerTxnRoot());
+
+    // No amplification needed, simply return existing transactions and results
+    if (n == 0)
+    {
+        transactions.emplace_back(*mTransactionIter);
+        results.emplace_back(*mResultIter);
+        return mTransactionIter->v0().tx.operations.size();
+    }
+
+    auto const& env = *mTransactionIter;
+    TransactionEnvelope newEnv = env;
+    Transaction& newTx = newEnv.tx;
+
+    // Keep track of accounts that need to sign
+    std::vector<SecretKey> keys;
+    newEnv.signatures.clear();
+
+    addSignerKeys(env.tx.sourceAccount, ltx, keys, env, n);
+    SimulationUtils::updateAccountID(newTx.sourceAccount, n);
+
+    for (auto& op : newTx.operations)
+    {
+        // Add signer keys where needed before simulating the operation
+        if (op.sourceAccount)
+        {
+            addSignerKeys(*op.sourceAccount, ltx, keys, env, n);
+        }
+
+        // Prior to protocol 10, it is possible that we might have just added a
```

Comment:
NB: I don't know if we care about simulation for older versions of the protocol (older than more than a year ago let's say), this adds complexity that we don't need.